pkg/chains/storage/oci/oci.go

```diff
@@ -179,10 +188,156 @@ func (b *Backend) Type() string {
+		b.logger.Infof("Image Ref #%d: %s\n", i, repository)
+		repositories = append(repositories, repository)
+	}
+	return repositories, nil
+}
+
+// retrieveEnvelopes retrieves the enveloppes containing the payloads and the signatures
+// for a specific OCI image.
+func (b *Backend) retrieveEnvelopes(imageRef string, opts config.StorageOpts) ([]ssdsse.Envelope, error) {
+	ctx := context.TODO()
+
+	// Parse the image reference.
+	ref, err := name.ParseReference(imageRef)
+	if err != nil {
+		return []ssdsse.Envelope{}, err
+	}
+
+	// Retrieve the verifier from the key.
+	pubKey, err := publicKeyFromKeyRef(ctx, opts.Key)
+	if err != nil {
+		return []ssdsse.Envelope{}, errors.Wrap(err, "loading public key")
+	}
+
+	// Retrieve the target repository.
+	sigRepo, err := cli.TargetRepositoryForImage(ref)
+	if err != nil {
+		return []ssdsse.Envelope{}, err
+	}
+
+	// Prepare the registry options.
+	remoteOpts := []remote.Option{
+		remote.WithAuthFromKeychain(authn.DefaultKeychain),
+		remote.WithContext(ctx),
+		remote.WithUserAgent("cosign/" + cli.VersionInfo().GitVersion),
+	}
+
+	// Prepare the CheckOpts.
+	co := &cosign.CheckOpts{
+		RegistryClientOpts:   remoteOpts,
+		SigVerifier:          dsse.WrapVerifier(pubKey),
+		SignatureRepo:        sigRepo,
+		VerifyBundle:         false,
+		SigTagSuffixOverride: cosign.AttestationTagSuffix,
+	}
+
+	// Verify the payload.
+	verified, err := cosign.Verify(ctx, ref, co)
+	if err != nil {
+		return []ssdsse.Envelope{}, err
+	}
+	b.logger.Infof("%d verified payloads found\n", len(verified))
+
+	// Collect the verified envelopes.
+	envelopes := []ssdsse.Envelope{}
+	for _, vp := range verified {
+		// Deserialize the verified payload.
+		envelope := ssdsse.Envelope{}
+		if err := json.Unmarshal(vp.Payload, &envelope); err != nil {
+			return []ssdsse.Envelope{}, fmt.Errorf("cannot decode the envelope: %s", err)
+		}
+		envelopes = append(envelopes, envelope)
+	}
+
+	return envelopes, nil
+}
+
+// extractEnvelopes extracts the enveloppes found in all the images produced by a
+// specific taskrun.
+func (b *Backend) extractEnvelopes(opts config.StorageOpts) ([]ssdsse.Envelope, error) {
+	imageRefs, err := b.retrieveImageReferences()
+	if err != nil {
+		return []ssdsse.Envelope{}, fmt.Errorf("cannot retrieve images from the taskrun: %s", err)
+	}
+	if len(imageRefs) < 1 {
+		return []ssdsse.Envelope{}, fmt.Errorf("there is no image to verify in this taskrun: %s", b.tr.Name)
+	}
+	envelopes := []ssdsse.Envelope{}
+	for _, imageRef := range imageRefs {
+		envs, err := b.retrieveEnvelopes(imageRef, opts)
+		if err != nil {
+			return []ssdsse.Envelope{}, err
+		}
+		envelopes = append(envelopes, envs...)
+		b.logger.Debugf("extractEnvelopes: %d envelopes retrieved\n", len(envelopes))
+	}
+	return envelopes, nil
+}
+
 func (b *Backend) RetrieveSignature(opts config.StorageOpts) (string, error) {
-	return "", fmt.Errorf("not implemented")
+	envelopes, err := b.extractEnvelopes(opts)
```

Comment:
I'm wondering if we can take more advantage of the `cosign/pkg/oci` package in the cosign repository here. Do you think something like this would work?

```go
	// given the TaskRun, retrieve the signature
	images := artifacts.ExtractOCIImagesFromResults(b.tr, b.logger)
	if len(images) != 1 {
		return "", errors.New("multiple images found, unsure which to provide signature for")
	}
	// get signature for a specific image
	ref, ok := images[0].(name.Digest)
	if !ok {
		return "", errors.New("error parsing image")
	}
	img, err := ociremote.SignedImage(ref)
	if err != nil {
		return "", err
	}
	sigImage, err := img.Signatures()
	if err != nil {
		return "", err
	}
	sigs, err := sigImage.Get()
	if err != nil {
		return "", err
	}
	for _, s := range sigs {
		if sig, err := s.Base64Signature(); err == nil {
			return sig, nil
		}
	}
	return "", fmt.Errorf("failed to find a signature")
```

and this should require a small change for getting the payload instead:

```
	for _, s := range sigs {
		if payload, err := s.Payload(); err == nil {
			return string(payload), nil
		}
	}
```


(as mentioned in the other comment, I do think we should change the function definition here, but this is just meant to be an example!)

Reply:
Oh sure! I'll give it a try 👍

Reply:
@priyawadhwa Where did you get the `ociremote` package from? From version 1.2.1, it looks like it cannot be imported (`use of internal package github.com/sigstore/cosign/internal/oci/remote not allowed`), and if I upgrade the dependency to 1.3.1, it breaks a bunch of stuff.